rand_xoshiro/src/splitmix64.rs

```diff
@@ -0,0 +1,144 @@
+use byteorder::{ByteOrder, LittleEndian};
+use rand_core::le::read_u64_into;
+use rand_core::impls::fill_bytes_via_next;
+use rand_core::{RngCore, SeedableRng, Error};
+
+/// A splitmix64 random number generator.
+///
+/// The splitmix algorithm is not suitable for cryptographic purposes, but is
+/// very fast and has a 64 bit state.
+///
+/// The algorithm used here is translated from [the `splitmix64.c`
+/// reference source code](http://xoshiro.di.unimi.it/splitmix64.c) by
+/// Sebastiano Vigna. For `next_u32`, a more efficient mixing function taken
+/// from [`dsiutils`](http://dsiutils.di.unimi.it/) is used.
+#[allow(missing_copy_implementations)]
+#[derive(Debug, Clone)]
+pub struct SplitMix64 {
+    x: u64,
+}
+
+impl SplitMix64 {
+    /// Seed a `SplitMix64` from a `u64`.
+    pub fn from_seed_u64(seed: u64) -> SplitMix64 {
+        let mut x = [0; 8];
+        LittleEndian::write_u64(&mut x, seed);
+        SplitMix64::from_seed(x)
+    }
+}
+
+const PHI: u64 = 0x9e3779b97f4a7c15;
+
+impl RngCore for SplitMix64 {
+    #[inline]
+    fn next_u32(&mut self) -> u32 {
+        self.x = self.x.wrapping_add(PHI);
+        let mut z = self.x;
+        // David Stafford's
+        // (http://zimbry.blogspot.com/2011/09/better-bit-mixing-improving-on.html)
+        // "Mix4" variant of the 64-bit finalizer in Austin Appleby's
+        // MurmurHash3 algorithm.
```

Comment:
Can you explain what the link is? This algorithm is SplitMix. Isn't the point of this implementation to reproduce output from the official version?

Reply:
AFAIK, the official version only supports generating `u64` natively. This is a more efficient variant for generating `u32`.